File: Frontend/services/api/rag.ts
/**
 * RAG (Retrieval-Augmented Generation) API Service
 * 
 * Provides AI-powered parenting guidance by querying indexed documents
 * and generating contextual answers with source attribution.
 */

import httpClient from '../http';

/**
 * Source chunk returned with each answer
 */
export interface SourceChunk {
  id: string;
  text: string;
  source: string;
}

/**
 * Request payload for RAG queries
 */
export interface RAGQueryRequest {
  question: string;
  top_k?: number; // Optional: 1-10, defaults to 3
}

/**
 * Response from RAG query endpoints
 */
export interface RAGQueryResponse {
  question: string;
  answer: string;
  chunks: SourceChunk[];
  prompt?: string;
  error?: string;
}

/**
 * RAG service status information
 */
export interface RAGStatus {
  ready: boolean;
  collection?: string;
  document_count?: number;
  chroma_dir?: string;
  model?: string;
  message?: string;
  error?: string;
}

/**
 * RAG health check response
 */
export interface RAGHealth {
  service: string;
  ready: boolean;
  status: string;
}

/**
 * RAG API Service
 */
export const ragAPI = {
  /**
   * Query the RAG system with a parenting question
   * 
   * @param query - Question and optional parameters
   * @returns Answer with source attribution
   * 
   * @example
   * ```typescript
   * const result = await ragAPI.query({
   *   question: "How can I support my child's emotional development?",
   *   top_k: 3
   * });
   * console.log(result.answer);
   * ```
   */
  query: async (query: RAGQueryRequest): Promise<RAGQueryResponse> => {
    const resp = await httpClient.post<RAGQueryResponse>('/rag/query', query);
    
    if (!resp.success || !resp.data) {
      throw new Error(resp.error || 'RAG query failed');
    }
    
    return resp.data;
  },

  /**
   * Query using GET request (convenient for simple queries)
   * 
   * @param question - The parenting question
   * @param top_k - Optional number of sources to retrieve (1-10)
   * @returns Answer with source attribution
   * 
   * @example
   * ```typescript
   * const result = await ragAPI.queryGet(
   *   "What are good bedtime routines?",
   *   3
   * );
   * ```
   */
  queryGet: async (question: string, top_k?: number): Promise<RAGQueryResponse> => {
    const params = new URLSearchParams({ question });
    if (top_k !== undefined) {
      params.append('top_k', top_k.toString());
    }
    
    const resp = await httpClient.get<RAGQueryResponse>(`/rag/query?${params.toString()}`);
    
    if (!resp.success || !resp.data) {
      throw new Error(resp.error || 'RAG query failed');
    }
    
    return resp.data;
  },

  /**
   * Get RAG service status and configuration
   * 
   * @returns Service status including document count and readiness
   * 
   * @example
   * ```typescript
   * const status = await ragAPI.getStatus();
   * if (status.ready) {
   *   console.log(`${status.document_count} documents available`);
   * }
   * ```
   */
  getStatus: async (): Promise<RAGStatus> => {
    const resp = await httpClient.get<RAGStatus>('/rag/status');
    
    if (!resp.success || !resp.data) {
      throw new Error(resp.error || 'Failed to get RAG status');
    }
    
    return resp.data;
  },

  /**
   * Quick health check for RAG service
   * 
   * @returns Basic health status
   * 
   * @example
   * ```typescript
   * const health = await ragAPI.healthCheck();
   * console.log(`RAG service: ${health.status}`);
   * ```
   */
  healthCheck: async (): Promise<RAGHealth> => {
    const resp = await httpClient.get<RAGHealth>('/rag/health');
    
    if (!resp.success || !resp.data) {
      throw new Error(resp.error || 'Health check failed');
    }
    
    return resp.data;
  },
};

/**
 * Example parenting questions that work well with the RAG system
 */
export const EXAMPLE_QUESTIONS = [
  "How can I support my child's emotional development?",
  "What are effective strategies for managing toddler tantrums?",
  "How do I talk to my teenager about difficult topics?",
  "What are age-appropriate chores for a 7-year-old?",
  "How can I help my child with test anxiety?",
  "What are signs of learning difficulties in children?",
  "How do I set healthy screen time boundaries?",
  "What are good bedtime routines for young children?",
  "How can I encourage my child to be more independent?",
  "What should I do if my child is being bullied?",
];

export default ragAPI;

File: Backend/app/routers/rag.py
"""RAG (Retrieval-Augmented Generation) endpoints for parenting guidance.

Provides question-answering based on indexed parenting articles and resources.
"""
from fastapi import APIRouter, HTTPException, Query
from pydantic import BaseModel, Field
from typing import Optional, List, Dict, Any
from app.services.rag import rag_service

router = APIRouter()


class RAGQuery(BaseModel):
    """Request body for RAG queries."""
    question: str = Field(..., min_length=5, max_length=500, description="The parenting question to answer")
    top_k: Optional[int] = Field(None, ge=1, le=10, description="Number of relevant sources to retrieve (1-10)")


class SourceChunk(BaseModel):
    """A retrieved source chunk."""
    id: str
    text: str
    source: str


class RAGResponse(BaseModel):
    """Response from RAG query."""
    question: str
    answer: str
    chunks: List[SourceChunk]
    prompt: Optional[str] = None
    error: Optional[str] = None


class RAGStatus(BaseModel):
    """RAG service status."""
    ready: bool
    collection: Optional[str] = None
    document_count: Optional[int] = None
    chroma_dir: Optional[str] = None
    model: Optional[str] = None
    message: Optional[str] = None
    error: Optional[str] = None


@router.post("/query", response_model=RAGResponse, summary="Ask a parenting question")
async def query_rag(query: RAGQuery) -> RAGResponse:
    """
    Query the RAG system with a parenting question.
    
    The system will:
    1. Retrieve relevant passages from indexed parenting articles
    2. Generate a comprehensive answer using an LLM
    3. Return the answer along with source references
    
    **Example questions:**
    - "How can I support my child's emotional development?"
    - "What are effective strategies for managing toddler tantrums?"
    - "How do I talk to my teenager about difficult topics?"
    """
    result = await rag_service.query(query.question, query.top_k)
    
    return RAGResponse(
        question=result["question"],
        answer=result["answer"],
        chunks=[
            SourceChunk(
                id=chunk["id"],
                text=chunk["text"],
                source=chunk["source"]
            )
            for chunk in result.get("chunks", [])
        ],
        prompt=result.get("prompt"),
        error=result.get("error")
    )


@router.get("/query", response_model=RAGResponse, summary="Ask a parenting question (GET)")
async def query_rag_get(
    question: str = Query(..., min_length=5, max_length=500, description="The parenting question to answer"),
    top_k: Optional[int] = Query(None, ge=1, le=10, description="Number of relevant sources to retrieve")
) -> RAGResponse:
    """
    Query the RAG system using GET request (convenient for simple queries).
    
    Same functionality as POST /query but accessible via URL parameters.
    """
    result = await rag_service.query(question, top_k)
    
    return RAGResponse(
        question=result["question"],
        answer=result["answer"],
        chunks=[
            SourceChunk(
                id=chunk["id"],
                text=chunk["text"],
                source=chunk["source"]
            )
            for chunk in result.get("chunks", [])
        ],
        prompt=result.get("prompt"),
        error=result.get("error")
    )


@router.get("/status", response_model=RAGStatus, summary="Get RAG service status")
async def get_rag_status() -> RAGStatus:
    """
    Get the current status of the RAG service.
    
    Returns information about:
    - Whether the service is ready
    - Vector database collection details
    - Number of indexed documents
    - LLM model being used
    
    Useful for health checks and debugging.
    """
    status = rag_service.get_status()
    return RAGStatus(**status)


@router.get("/health", summary="Quick health check")
async def health_check() -> Dict[str, Any]:
    """Quick health check endpoint for the RAG service."""
    return {
        "service": "rag",
        "ready": rag_service.ready,
        "status": "ok" if rag_service.ready else "unavailable"
    }
